lib/perfdata/elasticsearchdatastreamwriter.cpp

```diff
@@ -0,0 +1,878 @@
+
+		String label = NormalizeElasticsearchFieldName(pdv->GetLabel());
+		pdFields->Set(label, metric);
+	}
+
+	return pdFields;
+}
+
+// user-defined tags, as specified in the ECS specification: https://www.elastic.co/docs/reference/ecs/ecs-base#field-tags
+static Array::Ptr ExtractTemplateTags(const MacroProcessor::ResolverList& resolvers,
+	const Array::Ptr& tagsTmpl, const Checkable::Ptr& checkable, const CheckResult::Ptr& cr)
+{
+	if (tagsTmpl == nullptr) {
+		return nullptr;
+	}
+
+	ObjectLock olock(tagsTmpl);
+	Array::Ptr tags = new Array();
+	for (const String tag : tagsTmpl) {
+		String missingMacro;
+		Value value = MacroProcessor::ResolveMacros(tag, resolvers, cr, &missingMacro);
+		if (!missingMacro.IsEmpty()) {
+			Log(LogDebug, "ElasticsearchDatastreamWriter")
+				<< "Missing macro for tag: " << tag
+				<< ". Missing: " << missingMacro
+				<< " for checkable '" << checkable->GetName() << "'. Skipping.";
+			continue;
+		}
+		tags->Add(value);
+	}
+
+	return tags;
+}
+
+// user-defined labels, as specified in the ECS specification: https://www.elastic.co/docs/reference/ecs/ecs-base#field-labels
+static Dictionary::Ptr ExtractTemplateLabels(const MacroProcessor::ResolverList& resolvers,
+	const Dictionary::Ptr& labelsTmpl, const Checkable::Ptr& checkable, const CheckResult::Ptr& cr)
+{
+	if (labelsTmpl == nullptr) {
+		return nullptr;
+	}
+
+	ObjectLock olock(labelsTmpl);
+	Dictionary::Ptr labels = new Dictionary();
+	for (const Dictionary::Pair& labelKv : labelsTmpl) {
+		String missingMacro;
+		Value value = MacroProcessor::ResolveMacros(labelKv.second, resolvers, cr, &missingMacro);
+		if (!missingMacro.IsEmpty()) {
+			Log(LogDebug, "ElasticsearchDatastreamWriter")
+				<< "Missing macro for label: " << labelKv.first
+				<< "Label: " << labelKv.second
+				<< ". Missing: " << missingMacro
+				<< " for checkable '" << checkable->GetName() << "'. Skipping.";
+			continue;
+		}
+		labels->Set(labelKv.first, value);
+	}
+
+	return labels;
+}
+
+String ElasticsearchDatastreamWriter::ExtractDatastreamNamespace(const MacroProcessor::ResolverList& resolvers,
+	const Checkable::Ptr& checkable, const CheckResult::Ptr& cr)
+{
+	String namespaceTmpl = GetDatastreamNamespace();
+
+	String missingMacro;
+	Value value = MacroProcessor::ResolveMacros(namespaceTmpl, resolvers, cr, &missingMacro);
+	if (!missingMacro.IsEmpty() || value.IsEmpty()) {
+		Log(LogDebug, "ElasticsearchDatastreamWriter")
+			<< "Missing value for namespace. Missing: " << missingMacro
+			<< " for checkable '" << checkable->GetName() << "'. Skipping.";
+		return "default";
+	}
+
+	return NormalizeElasticsearchFieldName(value);
+}
+
+bool ElasticsearchDatastreamWriter::Filter(const Checkable::Ptr& checkable)
+{
+	if (!m_CompiledFilter)
+		return true;
+
+	auto [host, service] = GetHostService(checkable);
+
+	Namespace::Ptr frameNS = new Namespace();
+	frameNS->Set("host", host);
+	frameNS->Set("service", service);
+
+	ScriptFrame frame(true, frameNS);
```

Comment:
I think you should sandbox the script frame here to not allow any unsafe usages, like calling functions with sideeffects.